lib/grpc/client/resolver/dns.ex

```diff
@@ -17,38 +17,58 @@ defmodule GRPC.Client.Resolver.DNS do
     uri = URI.parse(target)
     host = uri.host || target
     port = uri.port || 50051
-    host_charlist = String.to_charlist(host)
-
-    case adapter().lookup(host_charlist, :a) do
-      {:ok, addresses} ->
-        addrs =
-          Enum.map(addresses, fn ip ->
-            %{address: ip |> :inet.ntoa() |> to_string(), port: port}
-          end)
-
-        case adapter().lookup(~c"_grpc_config." ++ host_charlist, :txt) do
-          {:ok, txt_records} ->
-            service_config_json =
-              txt_records
-              |> Enum.map(&List.to_string/1)
-              |> Enum.find_value(fn str ->
-                case String.split(str, "grpc_config=") do
-                  [_, json] -> json
-                  _ -> nil
-                end
-              end)
-
-            {:ok, %{addresses: addrs, service_config: ServiceConfig.parse(service_config_json)}}
-
-          _ ->
-            {:ok, %{addresses: addrs, service_config: nil}}
-        end
-
-      {:error, reason} ->
-        {:error, {:dns_error, reason}}
+
+    with {:ok, addresses} <- lookup_addresses(host) do
+      addrs =
+        Enum.map(addresses, fn ip ->
+          %{address: :inet.ntoa(ip) |> to_string(), port: port}
+        end)
+
+      case lookup_service_config(host) do
+        {:ok, txt_records} ->
+          service_config_json = extract_service_config(txt_records)
+
+          {:ok,
+           %{
+             addresses: addrs,
+             service_config: ServiceConfig.parse(service_config_json)
+           }}
+
+        :no_config ->
+          {:ok, %{addresses: addrs, service_config: nil}}
+
+        {:error, reason} ->
+          {:error, {:dns_error, reason}}
+      end
+    else
+      {:error, reason} -> {:error, {:dns_error, reason}}
+    end
+  end
+
+  defp lookup_addresses(host) do
+    adapter().lookup(host, :a)
+  end
+
+  defp lookup_service_config(host) do
+    case adapter().lookup(~c"_grpc_config." ++ host, :txt) do
```

Comment:
I think this needs fixing (and tests). The default impl will break on String.to_charlist

Reply:
Done!